Sectioned solidimple

mauplatteau opened the pull request:
Updates on the Sectiones Solid Horizontal
SectionedSolid and SectionedHorizontal have also been refactored since the last pull request.

*EDIT:* Fixes #138 .

mauplatteau committed:
Added attributes 1-2 to sectioned solid

mauplatteau committed:
Merge branch 'development' into refactoringm

mauplatteau committed:
Update SolidModelConverter.h
IfcSectionesSolid has now been refactored

mauplatteau committed:
Added const to convertIfcSectionedSolid

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Update: Check dimensions, Give Directrix to curve converter and give CrossSections to Profile Converter

pjanck commented on Core/src/IfcGeometryConverter/SolidModelConverter.h:
```diff
@@ -174,9 +182,69 @@ namespace OpenInfraPlatform
+				//check dimensions and correct attributes sizes
+                if(vec_cross_sections.size() != vec_cross_section_positions.size())
+				{
+					std::cout << "CrossSections and CrossSectionsPositions are not equal in size " << std::endl;
+					return;
```
this should be an exception
```suggestion
					throw oip::InconsistentModellingException("CrossSections and CrossSectionsPositions are not equal in size.");
```

mauplatteau replied:
resolved in commit ccd8948

pjanck commented on Core/src/IfcGeometryConverter/SolidModelConverter.h:
```diff
@@ -174,9 +182,69 @@ namespace OpenInfraPlatform
+					std::cout << "CrossSections and CrossSectionsPositions are not equal in size " << std::endl;
+					return;
+				}
+				if ((vec_cross_sections.size() - vec_cross_section_positions.size()) > 1)
+				{
+					std::cout << "vec_cross_sections size decreased by one" << std::endl;
+					vec_cross_sections.pop_back();
+				}
+				else if ((vec_cross_section_positions.size() - vec_cross_sections.size()) > 1)
+				{
+					std::cout << "vec_cross_section_positions decreased by one" << std::endl;
+					vec_cross_section_positions.pop_back();
+				}
```
This will never be executed (even in the previous version of the code), since the `return` or `throw` make the flow leave the function. These are not needed.

mauplatteau replied:
ok, i changed the return for a oip::InconsistentModellingException. And if those two cases are not needed i will erase them as well in the code, i just thought they might be usefull.

pjanck replied:
No need for them, erasing is just.

mauplatteau replied:
resolved in commit ccd8948

pjanck commented on Core/src/IfcGeometryConverter/SolidModelConverter.h:
```diff
@@ -174,9 +182,69 @@ namespace OpenInfraPlatform
+	     		 std::vector<carve::geom::vector<3> > basis_curve_points;
+				 curveConverter->convertIfcCurve(directrix.lock(), basis_curve_points, segment_start_points);
+
+				 std::shared_ptr<carve::input::PolyhedronData> pipe_data(new carve::input::PolyhedronData());
```
It's not really a pipe, right? Perhaps a better name would be `body_data` or sth like that?

pjanck replied:
Please mark the conversation "resolved" when finished - it enables an oversight, what is finished and what not.

mauplatteau replied:
resolved in commit ccd8948

pjanck commented on Core/src/IfcGeometryConverter/SolidModelConverter.h:
```diff
@@ -174,9 +182,69 @@ namespace OpenInfraPlatform
+	     		 std::vector<carve::geom::vector<3> > basis_curve_points;
+				 curveConverter->convertIfcCurve(directrix.lock(), basis_curve_points, segment_start_points);
+
+				 std::shared_ptr<carve::input::PolyhedronData> pipe_data(new carve::input::PolyhedronData());
```
Also, a `std::shared_ptr<>` should be initialized with `std::make_shared<>`.

mauplatteau replied:
ok so like this then? std::shared_ptr<carve::input::PolyhedronData> body_data = std::make_shared<carve::input::PolyhedronData>();

mauplatteau replied:
resolved in commit ccd8948

pjanck commented on Core/src/IfcGeometryConverter/SolidModelConverter.h:
```diff
@@ -174,9 +182,69 @@ namespace OpenInfraPlatform
+				 // Less than two points is a point
+				 if (num_curve_points < 2)
+				 {
+					 std::cout << "IfcSectionedSolidHoritzontal: num curve points < 2" << std::endl;
+					 return;
```
Again, a `throw` would be more suitable.

mauplatteau replied:
resolved in commit ccd8948

pjanck commented on Core/src/IfcGeometryConverter/SolidModelConverter.h:
```diff
@@ -174,9 +182,69 @@ namespace OpenInfraPlatform
+		         }
+  
+			     // check if paths has been filled with the coordinates of the ProfileDef. if it is empty -> return;
+				 if (paths.size() == 0)
+				 {
+					 std::cout << "Profile converter could not find coordinates" << std::endl;
+					 return;
+				 }
```
These should be within the for loop above. Plus, it should throw an exception.

mauplatteau replied:
resolved in commit ccd8948

pjanck commented on Core/src/IfcGeometryConverter/SolidModelConverter.h:
```diff
@@ -174,9 +182,69 @@ namespace OpenInfraPlatform
+						 std::shared_ptr<ProfileConverterT<IfcEntityTypesT>> profile_converter = profileCache->getProfileConverter(vec_cross_sections[i]);
+						 profile_converter->simplifyPaths();
+						 //Save coordinates in paths
+						 const std::vector<std::vector<carve::geom::vector<2>>>& paths = profile_converter->getCoordinates();
```
You need a `std::vector<std::vector<std::vector<carve::geom::vector<2>>>>` defined outside of the for loop, and then always append the `getCoordinates` to it.

mauplatteau replied:
resolved in commit 4ac9204

mauplatteau committed:
Update SectionedSolidHoritzontal
changed return to Exception based on the feedback from today

pjanck commented on Core/src/IfcGeometryConverter/SolidModelConverter.h:
```diff
@@ -187,18 +187,7 @@ namespace OpenInfraPlatform
-				{
-					std::cout << "vec_cross_section_positions decreased by one" << std::endl;
-					vec_cross_section_positions.pop_back();
+					oip::InconsistentModellingException("CrossSections and CrossSectionsPositions are not equal in size.");
```
```suggestion
					throw oip::InconsistentModellingException(sectioned_solid_horizontal, "CrossSections and CrossSectionsPositions are not equal in size.");

```

mauplatteau replied:
resolved in commit f4786e8

pjanck commented on Core/src/IfcGeometryConverter/SolidModelConverter.h:
```diff
@@ -220,28 +209,25 @@ namespace OpenInfraPlatform
-					 return;
-				 }
-				 
+					 throw oip::InconsistentModellingException("IfcSectionedSolidHoritzontal: num curve points < 2");
```
```suggestion
					 throw oip::InconsistentModellingException(sectioned_solid_horizontal, "num curve points < 2");

```

mauplatteau replied:
resolved in commit f4786e8

pjanck commented on Core/src/IfcGeometryConverter/SolidModelConverter.h:
```diff
@@ -220,28 +209,25 @@ namespace OpenInfraPlatform
+					     //check if paths has been filled with the coordinates of the ProfileDef. if empty -> throw Exception
+						 if (paths.size() == 0)
+						 {
+							 throw oip::InconsistentModellingException("Profile converter could not find coordinates");
```
```suggestion
							 throw oip::InconsistentModellingException(sectioned_solid_horizontal, "Profile converter could not find coordinates");

```

mauplatteau replied:
resolved in commit f4786e8

pjanck reviewed (changes requested):
Please also correct the indentation of code. See other code for example how. (General rule = { -> increase indentation, } -> decrease indentation) This would make the code more readable.

mauplatteau committed:
Minor update to commit  ccd8948

mauplatteau committed:
Now CrossSections are saved correctly

mauplatteau committed:
Fixed commit 91351da now it should be correct

mauplatteau committed:
Refactored CrossSectionPositions

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Update: 1.Get offset from curve, 2.  calculate the position on and the direction of the base curve
3. Calculate rotations, 4. Calculate position: still working on 4

mauplatteau committed:
Update: CrossSectionPositions
1.pointsOnCurve and directionsOfCurve are now saved in a vector.
2.Started working on the Tesselation
3.erased some old code as well

mauplatteau committed:
Update: points of CrossSectionPositions and curveConverter are merged into a list

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Interpolation of the Direction for BasisCurvePoints

mauplatteau committed:
1.Updated interpolation of directions
2.started working on the interpolation of Profiles

mauplatteau committed:
Started adding Vertex

mauplatteau committed:
Update: 1.new Vector paths_for_tesselation in 3D for Profile points
2.Added Vertex
3. Working on Profile interpolation

pjanck commented:
Great work so far! It looks very promising :)

mauplatteau committed:
Update: interpolation of Profiles

mauplatteau committed:
Merge remote-tracking branch 'tumcms/visual-unit-tests' into SectionedSolidimple

mauplatteau committed:
Corrected the merge, now it compiles

mauplatteau committed:
Still working on the push_back of vec_cross_sections

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Front and Back Faces of the profile added. The Profile along the directrix should also work now.

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Corrected some out of Range loops and also the interpolation

mauplatteau committed:
Tried correcting Profiledef

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Corrected one iterator in the vertex

mauplatteau committed:
Fixed out of range addFace loop

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Working on correcting the interpolation and on Front and back profile, the rest should be working now.

mauplatteau committed:
Changed Interpolation, localPlacementMatrix still has to be interpolated.

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Defined new Matrix operations

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Update to the last commit, BasisCurvePoints have a direction.
No need for the Interpolation of the direction anymore.

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
All points are now only multiplied with object_placement_matrix
Funktion in line 189 now also takes fixes_axis_vertival as a parameter

mauplatteau committed:
Corrected the Transformation Matrix
Closed the front and Back Profiles
Working on closing the Body

mauplatteau committed:
Closed the body of the geometric form

mauplatteau committed:
Corrected the last triangle in the body

pjanck commented:
This is starting to look pretty much finished. Can we convert this Pull Request from draft, is it ready for final review?

mauplatteau committed:
ClosingBody is now working as it should

pjanck reviewed (commented):
Can we slowly merge this work?

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Corrected typos

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Merge remote-tracking branch 'origin/SectionedSolidimple' into SectionedSolidimple

pjanck reviewed (commented):
As discussed:
- Unit tests should be submitted as well.
- The mistake in the angle calculation should be found.

mauplatteau committed:
working on UnitTest for SSH examples

mauplatteau committed:
Now the UnitTest compiles

mauplatteau commented:
The UnitTest is compiling without errors but i still can not find the pictures

mauplatteau committed:
Changed Classnaame to IFCSECTIONEDSOLIDHORIZONTAL

mauplatteau committed:
SSH-Rectangle_Var UnitTest

mauplatteau committed:
UnitTest with SSH on a line segment are working

pjanck reviewed (commented):
This looks very nice. Good job!

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Added UnitTests for the examples on an IfcAlignmentCurve

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Saved new images from UnitTest for 3 Examples

pjanck reviewed (commented):
Great figures, looks neat.

Now the only thing that remains, is the angle. I will send you some additional files to play with, perhaps that may help determine the mistake. I would also like you to include them (if/when they are working) as unit tests as well.

mauplatteau committed:
Merge branch 'development' into SectionedSolidimple

mauplatteau committed:
Merge remote-tracking branch 'tumcms/development' into SectionedSolidimple

mauplatteau committed:
Added the two missing Curve UnitTests

pjanck committed:
Merge branch 'development' into SectionedSolidimple
